File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/util/ViewModelFactory.java
package io.github.skywalkerdarren.simpleaccounting.util;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import io.github.skywalkerdarren.simpleaccounting.model.AppRepository;
import io.github.skywalkerdarren.simpleaccounting.view_model.AccountViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.BillDetailViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.BillEditViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.BillListViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.ChartViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.ClassifyViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.DiscoveryViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.EmptyListViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.FeedBackViewModel;
import io.github.skywalkerdarren.simpleaccounting.view_model.JournalViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private static volatile ViewModelFactory INSTANCE;
    private final AppRepository mRepository;

    private ViewModelFactory(AppRepository repository) {
        mRepository = repository;
    }

    public static ViewModelFactory getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory(AppRepository.getInstance(new AppExecutors(), application));
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    public static void destroyInstance() {
        INSTANCE = null;
    }

    public AppRepository getRepository() {
        return mRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(AccountViewModel.class)) {
            //noinspection unchecked
            return (T) new AccountViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(BillDetailViewModel.class)) {
            //noinspection unchecked
            return (T) new BillDetailViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(BillEditViewModel.class)) {
            //noinspection unchecked
            return (T) new BillEditViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(BillListViewModel.class)) {
            //noinspection unchecked
            return (T) new BillListViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(ClassifyViewModel.class)) {
            //noinspection unchecked
            return (T) new ClassifyViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(JournalViewModel.class)) {
            //noinspection unchecked
            return (T) new JournalViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(ChartViewModel.class)) {
            //noinspection unchecked
            return (T) new ChartViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(DiscoveryViewModel.class)) {
            //noinspection unchecked
            return (T) new DiscoveryViewModel(mRepository);
        } else if (modelClass.isAssignableFrom(EmptyListViewModel.class)) {
            //noinspection unchecked
            return (T) new EmptyListViewModel();
        } else if (modelClass.isAssignableFrom(FeedBackViewModel.class)) {
            //noinspection unchecked
            return (T) new FeedBackViewModel();
        }

        throw new IllegalArgumentException("no this ViewModel");
    }
}

File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/util/ViewModelFactory.kt
package io.github.skywalkerdarren.simpleaccounting.util

import android.app.Application
import androidx.annotation.VisibleForTesting
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import androidx.lifecycle.ViewModelProviders
import io.github.skywalkerdarren.simpleaccounting.model.AppRepository
import io.github.skywalkerdarren.simpleaccounting.model.repository.CurrencyRepo
import io.github.skywalkerdarren.simpleaccounting.view_model.*

class ViewModelFactory private constructor(private val mRepository: AppRepository, private val mCurrencyRepo: CurrencyRepo) : NewInstanceFactory() {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        when {
            modelClass.isAssignableFrom(AccountViewModel::class.java) -> {
                return AccountViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(BillDetailViewModel::class.java) -> {
                return BillDetailViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(BillEditViewModel::class.java) -> {
                return BillEditViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(BillListViewModel::class.java) -> {
                return BillListViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(ClassifyViewModel::class.java) -> {
                return ClassifyViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(JournalViewModel::class.java) -> {
                return JournalViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(ChartViewModel::class.java) -> {
                return ChartViewModel(mRepository) as T
            }
            modelClass.isAssignableFrom(DiscoveryViewModel::class.java) -> {
                return DiscoveryViewModel(mRepository, mCurrencyRepo) as T
            }
            modelClass.isAssignableFrom(EmptyListViewModel::class.java) -> {
                return EmptyListViewModel() as T
            }
            modelClass.isAssignableFrom(FeedBackViewModel::class.java) -> {
                return FeedBackViewModel() as T
            }
            modelClass.isAssignableFrom(CurrencyFavViewModel::class.java) -> {
                return CurrencyFavViewModel(mCurrencyRepo) as T
            }
            modelClass.isAssignableFrom(CurrencySelectViewModel::class.java) -> {
                return CurrencySelectViewModel(mCurrencyRepo) as T
            }
            else -> throw IllegalArgumentException("no this ViewModel")
        }
    }

    fun <T : ViewModel> obtainViewModel(fragment: Fragment, modelClass: Class<T>) =
            ViewModelProviders.of(fragment, this).get(modelClass)

    companion object {
        @Volatile
        private var INSTANCE: ViewModelFactory? = null

        @JvmStatic
        fun getInstance(application: Application): ViewModelFactory = INSTANCE
                ?: synchronized(this) {
                    INSTANCE ?: ViewModelFactory(
                            InjectorUtils.getAppRepo(application),
                            InjectorUtils.getCurrencyRepo(application)
                    ).also { INSTANCE = it }
                }

        @VisibleForTesting
        fun destroyInstance() {
            INSTANCE = null
        }
    }

}